File: artifacts/api-server/routes/registrations.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional
from database import get_db
import models
from auth_utils import get_current_user, require_roles

router = APIRouter()

def fmt_user(u):
    if not u:
        return None
    return {
        "id": u.id, "name": u.name, "email": u.email,
        "role": u.role, "department": u.department,
        "createdAt": u.created_at.isoformat() if u.created_at else "",
    }

def fmt_event(e):
    if not e:
        return None
    return {
        "id": e.id, "title": e.title, "description": e.description,
        "date": e.date, "time": e.time, "venue": e.venue,
        "department": e.department, "imageUrl": e.image_url,
        "rules": e.rules, "registrationFee": e.registration_fee,
        "totalSlots": e.total_slots, "availableSlots": e.available_slots,
        "paymentQrUrl": e.payment_qr_url, "status": e.status,
        "createdByFacultyId": e.created_by_faculty_id, "createdByFacultyName": None,
        "createdAt": e.created_at.isoformat() if e.created_at else "",
    }

def fmt_reg(r, event=None, student=None):
    return {
        "id": r.id, "eventId": r.event_id, "studentId": r.student_id,
        "paymentScreenshotUrl": r.payment_screenshot_url,
        "paymentStatus": r.payment_status,
        "adminComment": r.admin_comment,
        "registeredAt": r.registered_at.isoformat() if r.registered_at else "",
        "event": fmt_event(event),
        "student": fmt_user(student),
    }

class CreateRegistrationBody(BaseModel):
    eventId: int
    paymentScreenshotUrl: Optional[str] = None

@router.post("/registrations", status_code=201)
def create_registration(
    body: CreateRegistrationBody,
    current_user: models.User = Depends(require_roles("student")),
    db: Session = Depends(get_db),
):
    event = db.query(models.Event).filter(models.Event.id == body.eventId).first()
    if not event:
        raise HTTPException(404, detail="Event not found")
    if event.status != "approved":
        raise HTTPException(400, detail="Event is not open for registration")
    if event.available_slots <= 0:
        raise HTTPException(400, detail="No slots available")

    existing = db.query(models.Registration).filter(
        models.Registration.event_id == body.eventId,
        models.Registration.student_id == current_user.id,
    ).first()
    if existing:
        raise HTTPException(400, detail="Already registered for this event")

    reg = models.Registration(
        event_id=body.eventId,
        student_id=current_user.id,
        payment_screenshot_url=body.paymentScreenshotUrl,
        payment_status="pending",
    )
    db.add(reg)
    event.available_slots -= 1
    db.commit()
    db.refresh(reg)
    db.refresh(event)
    return fmt_reg(reg, event)

@router.get("/registrations/my")
def my_registrations(
    current_user: models.User = Depends(require_roles("student")),
    db: Session = Depends(get_db),
):
    regs = db.query(models.Registration).filter(
        models.Registration.student_id == current_user.id
    ).all()
    all_events = {e.id: e for e in db.query(models.Event).all()}
    return [fmt_reg(r, all_events.get(r.event_id)) for r in regs]

class ReviewBody(BaseModel):
    paymentStatus: str
    adminComment: Optional[str] = None

@router.patch("/registrations/{reg_id}/review")
def review_registration(
    reg_id: int,
    body: ReviewBody,
    current_user: models.User = Depends(require_roles("admin", "faculty")),
    db: Session = Depends(get_db),
):
    reg = db.query(models.Registration).filter(models.Registration.id == reg_id).first()
    if not reg:
        raise HTTPException(404, detail="Registration not found")

    old_status = reg.payment_status
    reg.payment_status = body.paymentStatus
    reg.admin_comment = body.adminComment

    if body.paymentStatus == "rejected" and old_status != "rejected":
        event = db.query(models.Event).filter(models.Event.id == reg.event_id).first()
        if event:
            event.available_slots += 1

    db.commit()
    db.refresh(reg)
    event = db.query(models.Event).filter(models.Event.id == reg.event_id).first()
    student = db.query(models.User).filter(models.User.id == reg.student_id).first()
    return fmt_reg(reg, event, student)

File: artifacts/api-server/src/routes/registrations.ts
import { Router } from "express";
import { db, registrationsTable, eventsTable, usersTable } from "@workspace/db";
import { eq, and } from "drizzle-orm";
import { requireAuth, requireRole } from "../middlewares/auth";
import { CreateRegistrationBody, ReviewRegistrationParams, ReviewRegistrationBody } from "@workspace/api-zod";
import { logger } from "../lib/logger";

const router = Router();

function formatReg(
  r: typeof registrationsTable.$inferSelect,
  event?: typeof eventsTable.$inferSelect | null,
  student?: typeof usersTable.$inferSelect | null
) {
  const formatted: Record<string, unknown> = {
    id: r.id,
    eventId: r.eventId,
    studentId: r.studentId,
    paymentScreenshotUrl: r.paymentScreenshotUrl ?? null,
    paymentStatus: r.paymentStatus,
    adminComment: r.adminComment ?? null,
    registeredAt: r.registeredAt.toISOString(),
    event: null,
    student: null,
  };
  if (event) {
    formatted.event = {
      id: event.id,
      title: event.title,
      description: event.description,
      date: event.date,
      time: event.time,
      venue: event.venue,
      department: event.department,
      imageUrl: event.imageUrl ?? null,
      rules: event.rules ?? null,
      registrationFee: event.registrationFee,
      totalSlots: event.totalSlots,
      availableSlots: event.availableSlots,
      paymentQrUrl: event.paymentQrUrl ?? null,
      status: event.status,
      createdByFacultyId: event.createdByFacultyId ?? null,
      createdByFacultyName: null,
      createdAt: event.createdAt.toISOString(),
    };
  }
  if (student) {
    formatted.student = {
      id: student.id,
      name: student.name,
      email: student.email,
      role: student.role,
      department: student.department,
      createdAt: student.createdAt.toISOString(),
    };
  }
  return formatted;
}

router.post("/registrations", requireAuth, requireRole("student"), async (req, res): Promise<void> => {
  const parsed = CreateRegistrationBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const [event] = await db.select().from(eventsTable).where(eq(eventsTable.id, parsed.data.eventId));
  if (!event) {
    res.status(404).json({ error: "Event not found" });
    return;
  }
  if (event.status !== "approved") {
    res.status(400).json({ error: "Event is not open for registration" });
    return;
  }
  if (event.availableSlots <= 0) {
    res.status(400).json({ error: "No slots available" });
    return;
  }

  const existing = await db.select().from(registrationsTable).where(
    and(eq(registrationsTable.eventId, parsed.data.eventId), eq(registrationsTable.studentId, req.user!.userId))
  );
  if (existing.length > 0) {
    res.status(400).json({ error: "Already registered for this event" });
    return;
  }

  const [reg] = await db.insert(registrationsTable).values({
    eventId: parsed.data.eventId,
    studentId: req.user!.userId,
    paymentScreenshotUrl: parsed.data.paymentScreenshotUrl ?? null,
    paymentStatus: "pending",
  }).returning();

  // Decrement slot immediately on registration so overbooking is prevented
  await db.update(eventsTable)
    .set({ availableSlots: event.availableSlots - 1 })
    .where(eq(eventsTable.id, event.id));

  const [updatedEvent] = await db.select().from(eventsTable).where(eq(eventsTable.id, event.id));
  res.status(201).json(formatReg(reg, updatedEvent));
});

router.get("/registrations/my", requireAuth, requireRole("student"), async (req, res): Promise<void> => {
  const regs = await db.select().from(registrationsTable).where(eq(registrationsTable.studentId, req.user!.userId));
  const eventIds = [...new Set(regs.map((r) => r.eventId))];
  const events: Record<number, typeof eventsTable.$inferSelect> = {};
  if (eventIds.length > 0) {
    const evts = await db.select().from(eventsTable);
    for (const e of evts) events[e.id] = e;
  }
  res.json(regs.map((r) => formatReg(r, events[r.eventId] ?? null)));
});

router.patch("/registrations/:id/review", requireAuth, requireRole("admin", "faculty"), async (req, res): Promise<void> => {
  const params = ReviewRegistrationParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: "Invalid registration ID" });
    return;
  }
  const parsed = ReviewRegistrationBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const [existing] = await db.select().from(registrationsTable).where(eq(registrationsTable.id, params.data.id));
  if (!existing) {
    res.status(404).json({ error: "Registration not found" });
    return;
  }

  const [reg] = await db.update(registrationsTable).set({
    paymentStatus: parsed.data.paymentStatus,
    adminComment: parsed.data.adminComment ?? null,
  }).where(eq(registrationsTable.id, params.data.id)).returning();

  // If rejected, give the slot back so another student can register
  if (parsed.data.paymentStatus === "rejected" && existing.paymentStatus !== "rejected") {
    const [evt] = await db.select().from(eventsTable).where(eq(eventsTable.id, reg.eventId));
    if (evt) {
      await db.update(eventsTable)
        .set({ availableSlots: evt.availableSlots + 1 })
        .where(eq(eventsTable.id, evt.id));
    }
  }

  logger.info({ registrationId: reg.id, status: parsed.data.paymentStatus }, "Registration reviewed");

  const [event] = await db.select().from(eventsTable).where(eq(eventsTable.id, reg.eventId));
  const [student] = await db.select().from(usersTable).where(eq(usersTable.id, reg.studentId));
  res.json(formatReg(reg, event, student));
});

export default router;
